network/src/topology.rs

```diff
@@ -0,0 +1,492 @@
+#[derive(Debug, Eq, Copy, Clone)]
+pub struct Connection {
+    /// One side of the connection.
+    pub source: SocketAddr,
+    /// The other side of the connection.
+    pub target: SocketAddr,
+    /// The last time this peer was seen by the crawler (used determine which connections are
+    /// likely stale).
+    last_seen: DateTime<Utc>,
+}
+
+impl PartialEq for Connection {
+    fn eq(&self, other: &Self) -> bool {
+        let (a, b) = (self.source, self.target);
+        let (c, d) = (other.source, other.target);
+
+        a == d && b == c || a == c && b == d
+    }
+}
+
+impl Hash for Connection {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let (a, b) = (self.source, self.target);
+
+        match a.cmp(&b) {
+            Ordering::Greater => {
+                b.hash(state);
+                a.hash(state);
+            }
+
+            _ => {
+                a.hash(state);
+                b.hash(state);
+            }
+        }
+    }
+}
+
+impl Connection {
+    fn new(source: SocketAddr, target: SocketAddr) -> Self {
+        Connection {
+            source,
+            target,
+            last_seen: Utc::now(),
+        }
+    }
+}
+
+/// Keeps track of crawled peers and their connections.
+#[derive(Debug, Default)]
+pub struct NetworkTopology {
+    connections: RwLock<HashSet<Connection>>,
+}
+
+impl NetworkTopology {
+    /// Updates the crawled connection set.
+    pub fn update(&self, source: SocketAddr, peers: Vec<SocketAddr>) {
+        // Rules:
+        //  - if a connecton exists already, do nothing.
+        //  - if a connection is new, add it.
+        //  - if an exisitng connection involving the source isn't in the peerlist, remove it if
+        //  it's stale.
+
+        let new_connections: HashSet<Connection> =
+            peers.into_iter().map(|peer| Connection::new(source, peer)).collect();
+
+        // Find which connections need to be removed.
+        //
+        // With sets: a - b = removed connections (if and only if one of the two addrs is the
+        // source), otherwise it's a connection which doesn't include the source and shouldn't be
+        // removed. We also keep connections seen within the last few hours as peerlists are capped
+        // in size and omitted connections don't necessarily mean they don't exist anymore.
+        let connections_to_remove: HashSet<Connection> = self
+            .connections
+            .read()
+            .difference(&new_connections)
+            .filter(|conn| {
+                (conn.source == source || conn.target == source)
+                    && (Utc::now() - conn.last_seen).num_hours() > STALE_CUTOFF
+            })
+            .copied()
+            .collect();
+
+        // Only retain connections that aren't removed.
+        self.connections
+            .write()
+            .retain(|connection| !connections_to_remove.contains(&connection));
+
+        // Scope the write lock.
```

Comment:
I'm guessing it's for future-proofing? (since the lock gets dropped right after this scope anyways)

Reply:
Indeed 👍